File: pubnub-java/src/main/java/com/pubnub/api/endpoints/remoteaction/RemoteAction.java
package com.pubnub.api.endpoints.remoteaction;

import com.pubnub.api.PubNubException;
import com.pubnub.api.callbacks.PNCallback;
import org.jetbrains.annotations.NotNull;

public interface RemoteAction<Output> {
    Output sync() throws PubNubException;

    void async(@NotNull PNCallback<Output> callback);

    void retry();

    void silentCancel();
}

File: core/src/main/kotlin/com/pubnub/api/endpoints/remoteaction/RemoteAction.kt
package com.pubnub.api.endpoints.remoteaction

import com.pubnub.api.PubNubException
import com.pubnub.api.callbacks.PNCallback
import com.pubnub.api.enums.PNOperationType

interface ExtendedRemoteAction<Output> : RemoteAction<Output> {
    fun operationType(): PNOperationType
    override fun retry()
}

interface RemoteAction<Output> : Cancelable {
    @Throws(PubNubException::class)
    fun sync(): Output?
    fun async(callback: PNCallback<Output>)
    fun retry()
    override fun silentCancel()
}

interface Cancelable {
    fun silentCancel()
}
